File: zkpass-sdk/rust/zkpass/apps/zkpass-demo/src/data_holder.rs
use std::time::Instant;
//use serde_json::Value;
use zkpass_client::interface::{
    ZkPassClient, ZkPassProofGenerator
};
use crate::data_issuer::DataIssuer;
use crate::proof_verifier::ProofVerifier;

pub struct DataHolder;

impl DataHolder {
    pub async fn start(&self, data_file: &str, dvr_file: &str) {
        //
        //  Get the user data from the data issuer
        //
        let data_issuer = DataIssuer;
        let user_data_token = data_issuer.get_user_data_token(data_file);

        //
        //  Get the dvr from the verifier
        //
        let proof_verifier = ProofVerifier;
        let dvr_token = proof_verifier.get_dvr_token(dvr_file);

        let zkpass_service_url = String::from("https://playground-zkpass.ssi.id/proof");
        //let zkpass_service_url = String::from("http://localhost:10888/proof");
        println!("\n#### starting zkpass proof generation...");
        let start = Instant::now();

        //
        //  Data Holder's integration points with the zkpass-client SDK library
        //

        //
        // Step 1: Instantiate the zkpass_client object.
        //
        let zkpass_client = ZkPassClient;

        //
        // Step 2: Call the zkpass_client.generate_zk_pass_proof
        //         to get the zkpass_proof_token.
        //
        let zkpass_proof_token = zkpass_client
            .generate_zkpass_proof(
                &zkpass_service_url,
                &user_data_token,
                &dvr_token)
            .await.unwrap();

        let duration = start.elapsed();
        println!("#### generation completed [time={:?}]", duration);

        //
        //  Step 3: Send the zkpass_proof_token to the Proof Verifier 
        //          to get the proof verified and retrieve the query result.
        //
        let query_result = proof_verifier
            .verify_zkpass_proof(zkpass_proof_token.as_str());

        println!("the query result is {}", query_result)
    }
}

File: typescript/node-js/cli/src/dataHolder.ts
/*
 * dataHolder.ts
 *
 * Authors:
 *   NaufalFakhri (naufal.f.muhammad@gdplabs.id)
 * Created Date: November 27th 2023
 * -----
 * Last Modified: November 29th 2023
 * Modified By: NaufalFakhri (naufal.f.muhammad@gdplabs.id)
 * -----
 * Reviewers:
 *   Zulchaidir (zulchaidir@gdplabs.id)
 *   Nugraha Tejapermana (nugraha.tejapermana@gdplabs.id)
 * ---
 * References:
 *   NONE
 * ---
 * Copyright (c) 2023 PT Darta Media Indonesia. All rights reserved.
 */
import { ZkPassClient } from "@didpass/zkpass-client-ts";

export class DataHolder {
  public async getProofToken(
    userDataToken: string,
    dvrToken: string
  ): Promise<string> {
    const ZKPASS_SERVICE_URL: string = "https://playground-zkpass.ssi.id/proof";

    try {
      console.log("\n#### starting zkpass proof generation...");
      const start = Date.now();

      //
      //  Data Holder's integration points with the zkpass-client SDK library
      //

      //
      // Step 1: Instantiate the ZkPassClient object.
      //
      const zkpassClient = new ZkPassClient();

      //
      // Step 2: Call the zkpassClient.generateZkpassProof
      //         to get the zkpassProofToken.
      //
      const zkpassProofToken = await zkpassClient.generateZkpassProof(
        ZKPASS_SERVICE_URL,
        userDataToken,
        dvrToken
      );

      const duration = Date.now() - start;
      console.log(`#### generation completed [time=${duration}ms]`);

      return zkpassProofToken;
    } catch (error) {
      console.log("#### DataHolder: error");
      console.log({ error });
    }
  }
}
